File: ABC/ABC015/d.cpp
#include <iostream>
#include <algorithm>
using namespace std;

const int INF = 1000000;
int dp[55][55][5050];
int W, N, K;

int main() {
    cin >> W;
    cin >> N >> K;
    for (int i=0; i<=K; ++i) for (int j=0; j<=5000; ++j) dp[0][i][j] = INF;
    dp[0][0][0] = 0;

    for (int i=1; i<=N; ++i) {
        int A, B;
        cin >> A >> B;
        for (int j=0; j<=K; ++j) {
            for (int k=0; k<=5000; ++k) {
                if (j >= 1 && k >= B) dp[i][j][k] = min(dp[i-1][j][k], dp[i-1][j-1][k-B]+A);
                else dp[i][j][k] = dp[i-1][j][k];
            }
        }
    }
    int ans = 0;
    for (int i=0; i<=5000; ++i) {
        for (int j=0; j<=K; ++j) {
            if (dp[N][j][i] <= W) ans = max(ans, i);
        }
    }

    cout << ans << endl;

    return 0;
}

File: ABC/ABC015/d.py
import numpy as np
INF = 10**12

W = int(input())
N, K = map(int, input().split(" "))
AB = np.array([[int(i) for i in input().split(" ")] for _ in range(N)])
Bsum = np.sum(AB[:,1])

dp = np.zeros((N+1, Bsum+1, K+1))
dp[0] = INF
dp[0][0][0] = 0

for i, (a, b) in enumerate(AB):
    dp[i+1][:b] = dp[i][:b]
    dp[i+1][b:][:,0] = dp[i][b:][:,0]
    dp[i+1][b:][:,1:] = np.minimum(dp[i][b:][:,1:], dp[i][:-b][:,:K] + a)

ans = 0
for v, ws in enumerate(dp[N]):
    if min(ws) <= W:
        ans = max(ans, v)
        
print(ans)